File: backend/app/schemas/graph.py
"""
Схемы графа влияний для интерактивной визуализации.

Формат специально подобран под react-force-graph-2d:
  nodes: список узлов
  links: список рёбер (source/target по id узла)

Это позволит на фронте делать:
    <ForceGraph2D
      graphData={data}
      nodeLabel="name"
      linkLabel={l => `${l.weight} (${l.confidence})`}
      ...
    />
"""
from typing import Literal

from pydantic import BaseModel, ConfigDict


class GraphNode(BaseModel):
    """Узел графа — режиссёр."""

    id: int                          # entity_id персоны
    name: str                        # имя на запрошенном языке
    image: str | None = None         # фото
    group: Literal["director"] = "director"
    influences_count: int = 0        # сколько раз был источником
    influenced_by_count: int = 0     # сколько раз был целью
    is_center: bool = False          # центральный узел в /api/graph/director/{id}

    model_config = ConfigDict(from_attributes=True)


class GraphLink(BaseModel):
    """Ребро графа — связь A повлиял на B."""

    source: int                      # source_director_id (наш узел)
    target: int                      # target_director_id (наш узел)
    weight: int                      # 1-5
    confidence: float                # 0-1
    relation_note: str | None = None

    model_config = ConfigDict(from_attributes=True)


class GraphResponse(BaseModel):
    """Полный ответ — то что нужно react-force-graph."""

    nodes: list[GraphNode]
    links: list[GraphLink]
    center_id: int | None = None     # для /api/graph/director/{id} — id центра
    depth: int = 1                   # сколько шагов в глубину

File: front/src/api/graph.ts
import { apiClient } from './client';

export interface GraphNode {
  id: number;
  name: string;
  image?: string | null;
  group?: string;
  influences_count?: number;
  influenced_by_count?: number;
  is_center?: boolean;
}

export interface GraphLink {
  source: number;
  target: number;
  weight?: number;
}

export interface GraphResponse {
  nodes: GraphNode[];
  links: GraphLink[];
  center_id?: number | null;
  depth?: number;
}

export interface RadialPerson {
  id: number;
  name: string;
  image: string | null;
  weight: number;
  films_count: number;
}

export interface RadialRing1Node extends RadialPerson {
  ring2: RadialPerson[];
}

export interface RadialCenter {
  id: number;
  name: string;
  image: string | null;
}

export interface RadialResponse {
  center: RadialCenter;
  ring1: RadialRing1Node[];
  neighbors: RadialPerson[];
  ring1_count: number;
  ring2_per_node: number;
  top_n_requested: number;
}

export async function getRadialGraph(
  directorId: number,
  topN = 4,
  ring2N = 3,
  lang: 'ru' | 'en' = 'ru'
): Promise<RadialResponse> {
  const { data } = await apiClient.get<RadialResponse>(
    `/api/graph/director/${directorId}/radial`,
    {
      params: { top_n: topN, ring2_n: ring2N, lang },
    }
  );
  return data;
}

export async function getFullGraph(limit = 50): Promise<GraphResponse> {
  const { data } = await apiClient.get<GraphResponse>('/api/graph/full', {
    params: { limit },
  });
  return data;
}

export async function getDirectorGraph(
  directorId: number,
  depth = 2,
  maxNodes = 80
): Promise<GraphResponse> {
  const { data } = await apiClient.get<GraphResponse>(
    `/api/graph/director/${directorId}`,
    {
      params: { depth, lang: 'ru', max_nodes: maxNodes },
    }
  );
  return data;
}
